### PR TITLE
WT-4199 Fix log corruption errantly reported due to buffer overrun

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

**Comment by ddanderson** [comment]
@sueloverso, can you please review this fix?

File: src/log/log.c [hunk 1]
```diff
@@ -1904,9 +1904,9 @@ __log_has_hole(WT_SESSION_IMPL *session, WT_FH *fh, wt_off_t log_size,
 			 * hole.  We can do some simple data verification to
 			 * ensure that this part of the file is not corrupted.
 			 */
-			for (p = buf; remainder > 0;
-			     remainder -= (wt_off_t)rdlen, p += rdlen) {
-				rdlen = WT_MIN(WT_LOG_ALIGN, (size_t)remainder);
+			for (p = buf, buf_left = rdlen; buf_left > 0;
+			     buf_left -= rdlen, p += rdlen) {
+				rdlen = WT_MIN(WT_LOG_ALIGN, buf_left);
```
Comment (sueloverso) [inline src/log/log.c]: I've been reading through this and thinking about this and understand it better. Although it isn't changed in this diff, I'm thinking `WT_LOG_ALIGN` should be `log->allocsize`. That is the unit that log records are aligned on. So if someone is using a non-standard log alignment, it should use that. (Might be good to add a non-standard log allocation size in the corruption tests too.)

File: src/log/log.c [hunk 2]
```diff
@@ -1904,9 +1904,9 @@ __log_has_hole(WT_SESSION_IMPL *session, WT_FH *fh, wt_off_t log_size,
 			 * hole.  We can do some simple data verification to
 			 * ensure that this part of the file is not corrupted.
```
Comment (sueloverso) [inline src/log/log.c]: I suggest getting rid of the "We can do..." sentence here. This code right here is only walking within the buffer to find the non-zero part. We're not verifying here yet.

File: src/log/log.c [hunk 3]
```diff
@@ -1916,7 +1916,7 @@ __log_has_hole(WT_SESSION_IMPL *session, WT_FH *fh, wt_off_t log_size,
 			 * present.  This record may also be partially written.
```
Comment (sueloverso) [inline src/log/log.c]: I think this comment here needs more based on my new understanding of this code today. What we're NOT doing is determining whether the rest of the file is corrupted. What we ARE doing is looking at the first non-zero bit that made us decide we have a hole and determining if that first non-zero part is a valid log record header (i.e. it is a hole) or if the first non-zero part is corruption/flipped bits, etc. I suggest:
/*
 * A presumed log record begins here where the buffer becomes non-zero. If we have
 * enough of a log record present in the buffer, we either have a valid header or corruption.
 * Verify the header of this record to determine whether it is just a hole or corruption.
 */

**Comment by sueloverso** [comment]
LGTM merge whenever you're ready

**Comment by ddanderson** [comment]
Thanks, @sueloverso, I'll merge.

**Pull request merged**